aws/tags.go

```diff
@@ -75,3 +78,42 @@ func ec2TagSpecificationsFromMap(m map[string]interface{}, t string) []*ec2.TagS
 		},
 	}
 }
+
+// ec2TagSpecificationsFromKeyValueTags returns the tag specifications for the given KeyValueTags object and resource type.
+func ec2TagSpecificationsFromKeyValueTags(tags keyvaluetags.KeyValueTags, t string) []*ec2.TagSpecification {
+	if len(tags) == 0 {
+		return nil
+	}
+
+	return []*ec2.TagSpecification{
+		{
+			ResourceType: aws.String(t),
+			Tags:         tags.IgnoreAws().Ec2Tags(),
+		},
+	}
+}
+
+// SetTagsDiff sets the new plan difference with the result of
+// merging resource tags on to those defined at the provider-level;
+// returns an error if unsuccessful or if the resource tags are identical
+// to those configured at the provider-level to avoid non-empty plans
+// after resource READ operations as resource and provider-level tags
+// will be indistinguishable when returned from an AWS API.
+func SetTagsDiff(_ context.Context, diff *schema.ResourceDiff, meta interface{}) error {
+	defaultTagsConfig := meta.(*AWSClient).DefaultTagsConfig
+	ignoreTagsConfig := meta.(*AWSClient).IgnoreTagsConfig
+
+	resourceTags := keyvaluetags.New(diff.Get("tags").(map[string]interface{}))
+
+	if defaultTagsConfig.TagsEqual(resourceTags) {
+		return fmt.Errorf(`"tags" are identical to those in the "default_tags" configuration block of the provider: please de-duplicate and try again`)
+	}
+
+	allTags := defaultTagsConfig.MergeTags(resourceTags).IgnoreConfig(ignoreTagsConfig)
```

Comment:
to accommodate cases where both (default and ignore are configured). do you think this is a valid use case?

Reply:
I think configuring both is certainly valid as long as the tag keys do not overlap. For example, if I'm working in engineering and running Kubernetes I may want to do something like:

```terraform
provider "aws" {
  default_tags {
    tags = {
      CostCenter = "Engineering"
    }
  }

  ignore_tags {
    key_prefixes = ["kubernetes.io/"]
  }
}
```

Maybe we should error in the overlapping keys case, if we do not already. Does not need to be in the initial pull request though. 😄